In [ ]:
#importing necessary libraries
import pandas as pd
import numpy as np

#importing library for importing data
from google.colab import files
uploaded = files.upload()

#by using pandas library, reading the dataset 
df = pd.read_csv("Data_User_Modeling_Dataset.csv")
df.info()

Saving Data_User_Modeling_Dataset.csv to Data_User_Modeling_Dataset.csv
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 403 entries, 0 to 402
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   STG     403 non-null    float64
 1   SCG     403 non-null    float64
 2   STR     403 non-null    float64
 3   LPR     403 non-null    float64
 4   PEG     403 non-null    float64
 5   UNS     403 non-null    object 
dtypes: float64(5), object(1)
memory usage: 19.0+ KB


In [ ]:
# loading more libraries for Classification analysis
#for decision tree classification
from sklearn.tree import DecisionTreeClassifier 
# for spilitting data in training and testing
from sklearn.model_selection import train_test_split 
# for accuracy of the model
from sklearn import metrics 

In [ ]:
df.head(20)

STG   SCG   STR   LPR   PEG   UNS
0   0.00  0.00  0.00  0.00  0.00   Low
1   0.08  0.08  0.10  0.24  0.90  High
2   0.06  0.06  0.05  0.25  0.33   Low
3   0.10  0.10  0.15  0.65  0.30  High
4   0.08  0.08  0.08  0.98  0.24   Low
5   0.09  0.15  0.40  0.10  0.66  High
6   0.10  0.10  0.43  0.29  0.56  High
7   0.15  0.02  0.34  0.40  0.01   Low
8   0.20  0.14  0.35  0.72  0.25   Low
9   0.00  0.00  0.50  0.20  0.85  High
10  0.18  0.18  0.55  0.30  0.81  High
11  0.06  0.06  0.51  0.41  0.30   Low
12  0.10  0.10  0.52  0.78  0.34  High
13  0.10  0.10  0.70  0.15  0.90  High
14  0.20  0.20  0.70  0.30  0.60  High
15  0.12  0.12  0.75  0.35  0.80  High
16  0.05  0.07  0.70  0.01  0.05   Low
17  0.10  0.25  0.10  0.08  0.33   Low
18  0.15  0.32  0.05  0.27  0.29   Low
19  0.20  0.29  0.25  0.49  0.56  High

In [ ]:
#Checking the unique values in "UNS" Column
df['UNS'].unique()

array(['Low', 'High'], dtype=object)

# DATA Preprocessing

In [ ]:
#Determinining the UNS variable as a categorical variable
df2 = df.copy()
df2["UNS"] = df2["UNS"].astype('category')

In [ ]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 403 entries, 0 to 402
Data columns (total 6 columns):
 #   Column  Non-Null Count  Dtype   
---  ------  --------------  -----   
 0   STG     403 non-null    float64 
 1   SCG     403 non-null    float64 
 2   STR     403 non-null    float64 
 3   LPR     403 non-null    float64 
 4   PEG     403 non-null    float64 
 5   UNS     403 non-null    category
dtypes: category(1), float64(5)
memory usage: 16.4 KB


In [ ]:
#by using label encoding converting categorical values into binary number (AS THERE ARE ONLY 2 LABELS SO, WE CAN USE LABEL ENCODING)
df2["UNS"] = df2["UNS"].cat.codes
df2.head()

STG   SCG   STR   LPR   PEG  UNS
0  0.00  0.00  0.00  0.00  0.00    1
1  0.08  0.08  0.10  0.24  0.90    0
2  0.06  0.06  0.05  0.25  0.33    1
3  0.10  0.10  0.15  0.65  0.30    0
4  0.08  0.08  0.08  0.98  0.24    1

# Descriptive statistics

In [ ]:
# descriptive statistics of these variables by using describe function
df2.describe()

STG         SCG         STR         LPR         PEG         UNS
count  403.000000  403.000000  403.000000  403.000000  403.000000  403.000000
mean     0.353141    0.355940    0.457655    0.431342    0.456360    0.444169
std      0.212018    0.215531    0.246684    0.257545    0.266775    0.497491
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000
25%      0.200000    0.200000    0.265000    0.250000    0.250000    0.000000
50%      0.300000    0.300000    0.440000    0.330000    0.400000    0.000000
75%      0.480000    0.510000    0.680000    0.650000    0.660000    1.000000
max      0.990000    0.900000    0.950000    0.990000    0.990000    1.000000

# Classification analysis

In [ ]:
# Assign the values of the UNS variable to the variable “labels” 
y = df2.UNS
y

0      1
1      0
2      1
3      0
4      1
      ..
398    0
399    0
400    1
401    0
402    0
Name: UNS, Length: 403, dtype: int8

In [ ]:
# dropping "UNS" variable from the data frame.
x = df2.drop(['UNS'], axis= 1)
x

STG   SCG   STR   LPR   PEG
0    0.00  0.00  0.00  0.00  0.00
1    0.08  0.08  0.10  0.24  0.90
2    0.06  0.06  0.05  0.25  0.33
3    0.10  0.10  0.15  0.65  0.30
4    0.08  0.08  0.08  0.98  0.24
..    ...   ...   ...   ...   ...
398  0.90  0.78  0.62  0.32  0.89
399  0.85  0.82  0.66  0.83  0.83
400  0.56  0.60  0.77  0.13  0.32
401  0.66  0.68  0.81  0.57  0.57
402  0.68  0.64  0.79  0.97  0.24

[403 rows x 5 columns]

## Split dataset as 80% training and 20% test sets.

In [ ]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.20, random_state=42)

## Decision tree algorithm.

In [ ]:
# decision tree classifier
dt_clf = DecisionTreeClassifier()

# Fitting the training set
dt_clf = dt_clf.fit(X_train,y_train)

# predit the result from testing set
y_pred_dt = dt_clf.predict(X_test)

In [ ]:
print("Accuracy of the Decision Tree model:", round(metrics.accuracy_score(y_test, y_pred_dt),2))

Accuracy of the Decision Tree model: 0.96


## Evaluation using accuracy, precision, recall, and F1 score metrics.

In [ ]:
# Accuracy of the DT classification model
print("Accuracy of the Decision Tree model:", round(metrics.accuracy_score(y_test, y_pred_dt),2))

Accuracy of the Decision Tree model: 0.96


In [ ]:
# importing the library for precision, recall and f1-score metrics.
from sklearn.metrics import classification_report
DecisionTree_report = classification_report(y_test,  y_pred_dt)
print(DecisionTree_report)

              precision    recall  f1-score   support

           0       0.96      0.98      0.97        47
           1       0.97      0.94      0.96        34

    accuracy                           0.96        81
   macro avg       0.96      0.96      0.96        81
weighted avg       0.96      0.96      0.96        81



# Regression analysis

In [ ]:
df2

STG   SCG   STR   LPR   PEG  UNS
0    0.00  0.00  0.00  0.00  0.00    1
1    0.08  0.08  0.10  0.24  0.90    0
2    0.06  0.06  0.05  0.25  0.33    1
3    0.10  0.10  0.15  0.65  0.30    0
4    0.08  0.08  0.08  0.98  0.24    1
..    ...   ...   ...   ...   ...  ...
398  0.90  0.78  0.62  0.32  0.89    0
399  0.85  0.82  0.66  0.83  0.83    0
400  0.56  0.60  0.77  0.13  0.32    1
401  0.66  0.68  0.81  0.57  0.57    0
402  0.68  0.64  0.79  0.97  0.24    0

[403 rows x 6 columns]

In [ ]:
#Assigning the values of the PEG variable to the variable “labels” 
y2 = df2.PEG
#dropping "PEG" variable from the data frame.
x2 = df2.drop(['PEG'], axis=1)
y2

0      0.00
1      0.90
2      0.33
3      0.30
4      0.24
       ... 
398    0.89
399    0.83
400    0.32
401    0.57
402    0.24
Name: PEG, Length: 403, dtype: float64

In [ ]:
x2

STG   SCG   STR   LPR  UNS
0    0.00  0.00  0.00  0.00    1
1    0.08  0.08  0.10  0.24    0
2    0.06  0.06  0.05  0.25    1
3    0.10  0.10  0.15  0.65    0
4    0.08  0.08  0.08  0.98    1
..    ...   ...   ...   ...  ...
398  0.90  0.78  0.62  0.32    0
399  0.85  0.82  0.66  0.83    0
400  0.56  0.60  0.77  0.13    1
401  0.66  0.68  0.81  0.57    0
402  0.68  0.64  0.79  0.97    0

[403 rows x 5 columns]

In [ ]:
#spiiltting new dataset as 80% training and 20% test sets. 
X2_train, X2_test, y2_train, y2_test = train_test_split(x2, y2, test_size=0.20, random_state=42)

## linear multiple regression

In [ ]:
#Importing sklearn library for linear regression
from sklearn.linear_model import LinearRegression
ml_regression = LinearRegression() 
#fitting training set in the linear regression model 
ml_regression.fit(X2_train, y2_train)

LinearRegression()

In [ ]:
#Intercept and Coefficient for the multilinear equation
print("Intercept: ", ml_regression.intercept_)
print("Coefficients:")
list(zip(x2, ml_regression.coef_))

Intercept:  0.7029727764096902
Coefficients:


[('STG', 0.03835889203531134),
 ('SCG', 0.033995783726359086),
 ('STR', -0.012806946733676418),
 ('LPR', -0.1565383427979285),
 ('UNS', -0.44217169498082826)]

ML Regression Equation: Equation = 0.7029 + (0.03836 * STG) + (0.033996* SCG) + (-0.0128 * STR) + (-0.15654 * LPR) + (-0.44217 * UNS) + e

In [ ]:
# Prediction of result from the test set
y_pred_mlr= ml_regression.predict(X2_test)

## Evaluation of model by presenting MSE, RMSE, and R-square values.

In [ ]:
# Evaluation of the linear regression model
MSE_mlr = round(metrics.mean_squared_error(y2_test, y_pred_mlr),3)
RMSE_mlr = round(np.sqrt(metrics.mean_squared_error(y2_test, y_pred_mlr)),3)
MAE_mlr = round(metrics.mean_absolute_error(y2_test, y_pred_mlr),3)
print('MSE (Mean Square Error):', MSE_mlr)
print('RMSE (Root Mean Square Error):', RMSE_mlr)

print("R-square value:", round(metrics.r2_score(y2_test, y_pred_mlr),3))

print('MAE (Mean Absolute Error):', MAE_mlr)


MSE (Mean Square Error): 0.018
RMSE (Root Mean Square Error): 0.136
R-square value: 0.749
MAE (Mean Absolute Error): 0.112


# References

[1]	“pandas - Python Data Analysis Library.” https://pandas.pydata.org/ (accessed Apr. 12, 2022).

[2]	“scikit-learn: machine learning in Python — scikit-learn 1.0.2 documentation.” https://scikit-learn.org/stable/ (accessed Apr. 12, 2022).

[3]	“3.3. Metrics and scoring: quantifying the quality of predictions,” scikit-learn. https://scikit-learn/stable/modules/model_evaluation.html (accessed Apr. 12, 2022).

[4]	R. Python, “Linear Regression in Python – Real Python.” https://realpython.com/linear-regression-in-python/ (accessed Apr. 12, 2022).

[5]	J. Brownlee, “Metrics To Evaluate Machine Learning Algorithms in Python,” Machine Learning Mastery, May 24, 2016. https://machinelearningmastery.com/metrics-evaluate-machine-learning-algorithms-python/ (accessed Apr. 12, 2022).

[6]	“pandas.DataFrame.describe — pandas 1.4.2 documentation.” https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.describe.html (accessed Apr. 12, 2022).

[7]	“Python Decision Tree Classification Tutorial: Scikit-Learn DecisionTreeClassifier,” DataCamp Community, Dec. 28, 2018. https://www.datacamp.com/community/tutorials/decision-tree-classification-python (accessed Apr. 12, 2022).

[8]	“sklearn.model_selection.train_test_split,” scikit-learn. https://scikit-learn/stable/modules/generated/sklearn.model_selection.train_test_split.html (accessed Apr. 12, 2022).

[9]	“sklearn.tree.DecisionTreeClassifier,” scikit-learn. https://scikit-learn/stable/modules/generated/sklearn.tree.DecisionTreeClassifier.html (accessed Apr. 12, 2022).

[10]	“sklearn.linear_model.LinearRegression,” scikit-learn. https://scikit-learn/stable/modules/generated/sklearn.linear_model.LinearRegression.html (accessed Apr. 12, 2022).

